File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/solid/dip/InventoryPort.java
package pe.joedayz.microservicios.modulo01.solid.dip;

import pe.joedayz.microservicios.modulo01.ddd.shared.Quantity;
import pe.joedayz.microservicios.modulo01.ddd.shared.Sku;
import pe.joedayz.microservicios.modulo01.tenant.TenantId;

/**
 * DIP (Dependency Inversion): PUERTO hacia el inventario.
 *
 * <p>El servicio de pedidos depende de esta abstraccion, no de un cliente HTTP concreto
 * ni de JDBC. En test se enchufa un adaptador en memoria; en produccion, uno que llama
 * por REST/gRPC al microservicio de Inventory. El dominio nunca importa detalles.
 */
public interface InventoryPort {

    /** @return true si logro reservar la cantidad para ese SKU en ese tenant. */
    boolean reserve(TenantId tenantId, Sku sku, Quantity quantity);

    /** Compensacion: libera una reserva previa (usada por la Saga al fallar un paso). */
    void release(TenantId tenantId, Sku sku, Quantity quantity);
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Solid/Dip/InventoryPort.cs
using System.Collections.Concurrent;
using JoeDayz.Microservicios.Modulo01.Ddd.Shared;
using JoeDayz.Microservicios.Modulo01.Tenant;

namespace JoeDayz.Microservicios.Modulo01.Solid.Dip;

/// <summary>
/// DIP: el caso de uso depende de esta ABSTRACCION, no del adaptador concreto.
/// Manana el inventario puede ser gRPC, REST o una cola: el dominio no cambia.
/// </summary>
public interface IInventoryPort
{
    bool Reserve(TenantId tenantId, Sku sku, Quantity quantity);

    void Release(TenantId tenantId, Sku sku, Quantity quantity);
}

/// <summary>Adaptador en memoria del puerto de inventario (aislado por tenant).</summary>
public sealed class InMemoryInventoryAdapter : IInventoryPort
{
    private readonly ConcurrentDictionary<string, int> _stock = new();

    public void SetStock(TenantId tenantId, Sku sku, int units) => _stock[Key(tenantId, sku)] = units;

    public int Available(TenantId tenantId, Sku sku) => _stock.GetValueOrDefault(Key(tenantId, sku), 0);

    public bool Reserve(TenantId tenantId, Sku sku, Quantity quantity)
    {
        var key = Key(tenantId, sku);
        var current = _stock.GetValueOrDefault(key, 0);
        if (current < quantity.Value)
        {
            return false;
        }

        _stock[key] = current - quantity.Value;
        return true;
    }

    public void Release(TenantId tenantId, Sku sku, Quantity quantity) =>
        _stock.AddOrUpdate(Key(tenantId, sku), quantity.Value, (_, current) => current + quantity.Value);

    private static string Key(TenantId tenantId, Sku sku) => $"{tenantId.Value}::{sku.Value}";
}
